File: 2020/day5/day5.js
const fs = require('fs');

const lines = fs.readFileSync('day5/day5.txt', { encoding: 'utf-8' }).split('\n');

/**
 * I: array of strings
 * O:
 * C:
 * E:
 * if F then it is lower half
 * if B then it is upper half
 */


const getRow = (line, low, high) => {
  if (!line.length) {
    return low;
  }
  let mid = (low + high) / 2;

  if (line[0] === 'F' || line[0] === 'L') {
    high = mid;
  }
  if (line[0] === 'B' || line[0] === 'R') {
    low = mid;
  }
  return getRow(line.slice(1), low, high);
};

const totals = [];

lines.forEach(line => {
  const row = getRow(line.slice(0, 7), 0, 128);
  const seat = getRow(line.slice(7), 0, 8);
  const seatNum = (row * 8) + seat;
  totals.push(seatNum)
});

console.log(totals.sort((a, b) => b - a)[0]);

File: 2020/day5/day5.rb
require 'csv'
require 'pry'

def get_row(row, high, low) 
  if row.nil?
    return low
  end

  mid = (low + high) / 2

  if row[0] == 'F' || row[0] == 'L'
    high = mid
  end

  if row[0] == 'B' || row[0] == 'R'
    low = mid
  end
  return get_row(row[1..-1], high, low)
end

max_id = 0

CSV.foreach('day5.txt') do |line|
  line = line.join.split('')

  row = get_row(line[0..6], 128, 0)
  seat = get_row(line[7..-1], 8, 0)
  id = (row * 8) + seat
  max_id = id if id > max_id
end

puts max_id